runtime/altair/src/migrations.rs

```diff
@@ -166,3 +169,145 @@ mod asset_registry {
+		fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+			let total_issuance =
+				orml_tokens::TotalIssuance::<Runtime>::get(DEPRECATED_AUSD_CURRENCY_ID);
+			let entries: Vec<(AccountId, AccountData<Balance>)> =
+				orml_tokens::Accounts::<Runtime>::iter()
+					.filter(|(_, old_currency_id, _)| {
+						*old_currency_id == DEPRECATED_AUSD_CURRENCY_ID
+					})
+					.map(|(account, _, account_data)| (account, account_data))
+					.collect::<_>();
+
+			Ok(OldState {
+				total_issuance,
+				entries,
+			}
+			.encode())
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn post_upgrade(state: Vec<u8>) -> Result<(), &'static str> {
+			use crate::OrmlTokens;
+
+			let old_state = OldState::decode(&mut state.as_ref())
+				.map_err(|_| "Error decoding pre-upgrade state")?;
+
+			let new_total_issuance =
+				orml_tokens::TotalIssuance::<Runtime>::get(NEW_AUSD_CURRENCY_ID);
+
+			ensure!(
+				old_state.total_issuance == new_total_issuance,
+				"The old AUSD issuance differs from the new one"
+			);
+
+			for (account, account_data) in old_state.entries {
+				ensure!(
+					OrmlTokens::accounts(&account, NEW_AUSD_CURRENCY_ID) == account_data.clone(),
+					"The account data under the new AUSD Currency does NOT match the old one"
+				);
+			}
+
+			return Ok(());
+		}
+
+		fn on_runtime_upgrade() -> Weight {
+			use frame_support::traits::tokens::fungibles::Mutate;
+
+			let mut migrated_entries = 0;
+
+			// Burn all AUSD tokens under the old CurrencyId and mint them under the new one
+			orml_tokens::Accounts::<Runtime>::iter()
+				.filter(|(_, old_currency_id, _)| *old_currency_id == DEPRECATED_AUSD_CURRENCY_ID)
+				.for_each(|(account, _, account_data)| {
+					let balance = account_data.free;
+					// Burn the amount under the old, hardcoded CurrencyId
+					<orml_tokens::Pallet<Runtime> as Mutate<AccountId>>::burn_from(
+						DEPRECATED_AUSD_CURRENCY_ID,
+						&account,
+						balance,
+					)
+					.map_err(|e| {
+						log::error!(
+							"Failed to call burn_from({:?}, {:?}, {balance}): {:?}",
+							DEPRECATED_AUSD_CURRENCY_ID,
+							account,
+							e
+						)
+					})
+					.ok();
+					// // Now mint the amount under the new CurrencyID
+					<orml_tokens::Pallet<Runtime> as Mutate<AccountId>>::mint_into(
+						NEW_AUSD_CURRENCY_ID,
+						&account,
+						balance,
+					)
+					.map_err(|e| {
+						log::error!(
+							"Failed to mint_into burn_from({:?}, {:?}, {balance}): {:?}",
+							NEW_AUSD_CURRENCY_ID,
+							account,
+							e
+						)
+					})
+					.ok();
+
+					migrated_entries += 1;
+				});
+
+			// Approximate weight given for every entry migration there are two calls being
+			// made, so counting one read and one write per call. Given that we have very
+			// few entries to be migrated, the specific value isn't quite relevant and we
+			// could just set this to a high-enough value.
+			<Runtime as frame_system::Config>::DbWeight::get()
+				.reads_writes(migrated_entries * 2, migrated_entries * 2)
```

Comment:
AFAICS, it should be one read and two writes per entry right?

Reply:
`burn_from` reads `TotalIssuance::<T>::get(` and `Self::accounts(who, currency_id).free`, so that's at least 2 reads per entry. As written in the comment - correct me if I am wrong - it doesn't really matter; this applies to just a few entries and the main purpose of weights and fees is to avoid spamming which is not a risk with these migrations.

Reply:
> main purpose of weights and fees is to avoid spamming which is not a risk with these migrations.

For extrinsics, this is true. For migrations however, we should always strive to provide a worst-case weight which always overestimates. Otherwise we could end up with tuple of migrations which exceed the maximum block weight such that the block in which the runtime upgrade is applied cannot be produced by para collators and relay validators with hardware close to our benchmarks. As long as this block is not produced, collators will continue trying to produce the same block and get it validated by relay validators. Depending on the amount of overweight, the chain could stall until relay governance comes to the rescue.

Reply:
Thanks for the context! 💛  

> For migrations however, we should always strive to provide a worst-case weight which always overestimates

That's why I specified what seems to me the correct estimation of reads and rights.

Do you think there's more to do on this thread?

Reply:
The current estimation does not overestimate! It should be 5 reads and 4 writes per item as burning and minting each read and write both `TotalIssuance` as well as `Account`. The last read comes from iterating the `Accounts` storage map.

Reply:
Done